crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,158 @@
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+use crate::core::decoder::DecoderVarInt;
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "use_serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ByteBuf {
+    #[cfg_attr(feature = "use_serde", serde(with = "base64"))]
+    inner: Vec<u8>,
+}
+
+impl ByteBuf {
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn inner(&self) -> Vec<u8> {
+        self.inner.clone()
+    }
+}
+
+impl From<Vec<u8>> for ByteBuf {
+    fn from(bytes: Vec<u8>) -> Self {
+        ByteBuf { inner: bytes }
+    }
+}
+
+impl Decoder for ByteBuf {
+    fn decode<T>(&mut self, src: &mut T, _version: Version) -> Result<(), Error>
+    where
+        T: Buf,
+    {
+        let mut len: i64 = 0;
+        len.decode_varint(src)?;
+
+        if len < 1 {
+            return Ok(());
+        }
+
+        self.inner.extend_from_slice(&src.chunk());
```

Comment:
`chunk` is working for small vectors, but its actually intended for internal usage.
I'm still investigating on a better approach than walking and executing `get_u8` per byte.

Reply:
Replaced with:

```rust
self.inner
            .extend_from_slice(&src.copy_to_bytes(len as usize));
```

Which will consume the amount of bytes from `len`.